Distribute particulate bottom flux within column

mnlevy1981 opened the pull request:
This offers a different strategy to the same problem #377 is attempting to solve. Here, the particulate that hits the sea floor is redistributed as remineralization within the entire column (rather than just in the bottom layer, which is how `development` currently handles it, and without computing a bottom flux for each tracer, which is how the other PR handles it).

This is a draft because two key pieces are missing:

1. Providing a mechanism for the GCM to pass `bot_flux_to_tend(:)` instead of assuming it is `1 / delta_z(column_kmt)` at level `column_kmt` and 0 elsewhere
2. Updating the CISO driver to behave similarly

Note that this approach should be bit-for-bit in POP (which is not the case with the other PR due to how we handle denitrification) and does not require changes to the conservation diagnostics

Fixes #360

mnlevy1981 committed:
Distribute particulate bottom flux within column

First step towards allowing the GCM to dictate how MARBL distributes the
particulates that hit the sea floor; rather than adding (to_floor - sed_loss) /
dz to the bottom layer, we add (to_floor - sed_loss) * bot_flux_to_tend(:) to
the whole column. For now, bot_flux_to_tend is 0 everywhere except the bottom
layer (where it is 1/dz), but that will change in a future commit.

mnlevy1981 committed:
bot_flux_to_tend should be length km, not kmt

Also, taking advantage of code that is in an if (k == column_kmt) block, I
shortened several instances of "1:column_kmt" to "1:k"

mnlevy1981 committed:
Redistribute remin in ciso

Same as previous commit (hard-coding bot_flux_to_tend to be 0 everywhere except
the kmt layer, where it is 1/dz(kmt)), applied to CISO module.

mnlevy1981 committed:
Move declaration of bot_flux_to_tend(:)

It is now declared in marbl_interior_tendency_compute() and passed to
compute_particulate_terms() as well as the CISO routine. This more closely
matches what we want as a final API (the GCM passing the array to MARBL)

mnlevy1981 committed:
bot_flux_to_tend is now part of interface class

The GCM needs to set marbl_instance%bot_flux_to_tend prior to calling
marbl_instance%interior_tendency_compute(). This array is allocated to be
length num_levels during initialization, and is passed as an argument to
marbl_interior_tendency_compute().

This commit also updates the stand-alone driver so that the
call_compute_subroutines test still works.

mnlevy1981 commented:
As of 816c3e5, GCM sets `marbl_instance%bot_flux_to_tend(:)` before each call to `marbl_instance%interior_tendency_compute()` (and the CISO fix went in dc90124, which was tested in POP and is bit-for-bit with marbl0.39.3 tag)

mnlevy1981 committed:
Add conservation check for bot_flux_to_tend

Need sum(dz * bot_flux_to_tend) == 1 to conserve

mnlevy1981 committed:
Lowered sum(dz * bot_flux_to_tend) threshold

It looks like 1e-14 would be a reasonable threshold based on running in MOM,
but I went with 1e-12 just to have a little wiggle room

mnlevy1981 committed:
Increase max_bracket_grow_it

Changed from max_bracket_grow_it=3 to max_bracket_grow_it=5, which was
necessary to converge when MOM columns had vanishingly thin layers at the
bottom.

klindsay28 commented:
Do you know if we need the increase in `max_bracket_grow_it` with `MIX_BOUNDARY_TRACER_ALE=True`?

If we need to increase `max_bracket_grow_it`, I'd like to understand why, to avoid the increase papering over a problem without us knowing that.

mnlevy1981 commented:
> Do you know if we need the increase in `max_bracket_grow_it` with `MIX_BOUNDARY_TRACER_ALE=True`?
> 
> If we need to increase `max_bracket_grow_it`, I'd like to understand why, to avoid the increase papering over a problem without us knowing that.

That's a good question. Yesterday I mentioned not being able to run on `izumi`, but this morning I woke up and realized I didn't include the updated version of `MOM_tracer_diabatic.F90` in my sandbox, so it's possible the code was hanging due to the parentheses bug in `tracer_vertdiff()` rather than memory or compiler issues... so I'm trying one more time to run on the CGD machine and if it works, I'll see if I can lower `max_bracket_grow_it` again.

mnlevy1981 commented:
> this morning I woke up and realized I didn't include the updated version of `MOM_tracer_diabatic.F90` in my sandbox, so it's possible the code was hanging due to the parentheses bug in `tracer_vertdiff()` rather than memory or compiler issues...

I also forgot to copy all the inputdata files over, which might actually have been the problem... so I guess on Thursday I can try again?

mnlevy1981 commented:
> > Do you know if we need the increase in `max_bracket_grow_it` with `MIX_BOUNDARY_TRACER_ALE=True`?
> > If we need to increase `max_bracket_grow_it`, I'd like to understand why, to avoid the increase papering over a problem without us knowing that.
> 
> That's a good question.

`max_bracket_grow_it = 3` died in `/glade/scratch/mlevy/C1850MOMECO.087/run` despite `MIX_BOUNDARY_TRACER_ALE=True`

```
WARNING from PE   509: (Task 509) Message from (lon, lat) (  49.333,   6.777). Level: 65
581:WARNING from PE   509: (Task 509) MARBL WARNING (marbl_co2calc_mod:drtsafe): (marbl_co2calc_mod:drtsafe) it = 4
581:WARNING from PE   509: (Task 509) MARBL WARNING (marbl_co2calc_mod:drtsafe): (marbl_co2calc_mod:drtsafe) x1,f =  0.9175434E-008-0.9062212E-004
581:WARNING from PE   509: (Task 509) MARBL WARNING (marbl_co2calc_mod:drtsafe): (marbl_co2calc_mod:drtsafe) x2,f =  0.1454208E-004-0.2180923E-002
581:WARNING from PE   509: (Task 509) MARBL ERROR (marbl_co2calc_mod:drtsafe): bounding bracket for pH solution not found
581:WARNING from PE   509: (Task 509) MARBL ERROR (marbl_co2calc_mod:drtsafe): (marbl_co2calc_mod:drtsafe) dic =  0.2000841E+004
581:WARNING from PE   509: (Task 509) MARBL ERROR (marbl_co2calc_mod:drtsafe): (marbl_co2calc_mod:drtsafe) ta =  0.2403694E+004
581:WARNING from PE   509: (Task 509) MARBL ERROR (marbl_co2calc_mod:drtsafe): (marbl_co2calc_mod:drtsafe) pt =  0.4128276E+000
581:WARNING from PE   509: (Task 509) MARBL ERROR (marbl_co2calc_mod:drtsafe): (marbl_co2calc_mod:drtsafe) sit =  0.6506462E+001
581:WARNING from PE   509: (Task 509) MARBL ERROR (marbl_co2calc_mod:drtsafe): (marbl_co2calc_mod:drtsafe) temp =  0.2542597E+002
581:WARNING from PE   509: (Task 509) MARBL ERROR (marbl_co2calc_mod:drtsafe): (marbl_co2calc_mod:drtsafe) salt =  0.3584534E+002
581:WARNING from PE   509: (Task 509) MARBL ERROR (marbl_co2calc_mod:comp_htotal): Error reported from drtsafe
```

I was thinking about two separate tests:

1. try `max_bracket_grow_it = 4` (I know 5 works and 3 doesn't)
2. switch to `nstep` output and see what things look like right before the failure; at this point, all I can tell is that the error occurs in the first day though I wouldn't be surprised if it was in the first time step.

Before I get to that, I realized that I forgot to update the dummy MARBL interface to include `bot_flux_to_tend` on the interface class... I'd like to get the CI passing, though I'm having trouble getting a compute node on cheyenne to test the build right now

mnlevy1981 committed:
Return max_bracket_grow_it to 3

Also needed to increase the tolerance for 1 - sum(dz*bot_flux_to_tend) from
1e-12 to 1e-10 (testing with a new scheme for computing bot_flux_to_tend in
MOM6 was giving differences ~1e-11)

mnlevy1981 committed:
Better tolerance for bot_flux_to_tend*dz

Thanks to improvements in how MOM computes bot_flux_to_tend, we can tighten the
tolerance on how far from 1 the sum of that array times dz can be. Also added
additional column information to the log when the error is triggered.

mnlevy1981 committed:
Add bftt_dz_sum_thres to settings

Rather than hard-code the 1e-14 threshold for 1-sum(bot_flux_to_tend*dz), a new
parameter is introduced that uses 1e-14 as a default but allows users to change
it without needing to rebuild the executable.
